### PR TITLE
Adding the repo..

### PR DESCRIPTION
**Pull request by hrigner**
The first commit containing the basic stuff needed in the repo.

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,21 @@
+The MIT License
+
+Copyright (c) 2018-present QlikTech International AB
```
Comment (peol) [inline LICENSE]: 2019 :)

File: README.md [hunk 2]
```diff
@@ -1,2 +1,24 @@
 # hamus.js
-hamus.js is a collection of React hooks for sharing reusable functionality used when dealing with Qlik Association Engine objects.
+
+hamus.js is a collection of React hooks for sharing reusable functionality used when
+dealing with Qlik Association Engine objects.
```
Comment (peol) [inline README.md]: `Association` -> `Associative`
`objects` -> `projects`

File: docs/useModel.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# useModel
+
+Creates a session object from a definition. The created object is saved in a cache to prevent multiple calls with the same app and definition.
+[CreateSessionObject](https://core.qlik.com/services/qix-engine/apis/qix/doc/#createsessionobject) is the QIX API method called.
+
+In addition to React 16.8.1 or later, [react-use-promise](https://www.npmjs.com/package/react-use-promise) must be installed to use this hook.
```
Comment (peol) [inline docs/useModel.md]: For convenience, perhaps this can be installed as a dependency instead of peer dep? One less thing for the consumer to worry about (I believe it's only a couple of lines).

File: docs/useModel.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    <div>
+      {content}
+    </div>
+  );
+};
+```
+
+## Reference
+
+Input parameters are [`app`](https://core.qlik.com/services/qix-engine/apis/qix/doc/) and [`definition`](https://core.qlik.com/services/qix-engine/apis/qix/definitions/#genericobjectproperties).
```
Comment (peol) [inline docs/useModel.md]: Should we mention/link to enigma.js here, since that's what we require as param here?

File: docs/useLayout.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  );
+};
+```
+
+# Reference
+
+```jsx
+const [layout, layoutError] = useLayout(model);
+```
+where `model` is either a QIX API [Doc](https://core.qlik.com/services/qix-engine/apis/qix/doc/) or a [GenericObject](https://core.qlik.com/services/qix-engine/apis/qix/genericobject/).
```
Comment (peol) [inline docs/useLayout.md]: Perhaps mention that it's an enigma.js model of type doc/generic object, which can be fetched with `useModel()` hook?

File: example/app.jsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
+import React, { useMemo, useRef } from 'react';
+import enigma from 'enigma.js';
+import schema from 'enigma.js/schemas/12.170.2.json';
+import usePromise from 'react-use-promise';
+import useModel from '../src/hooks/useModel';
```
Comment (peol) [inline example/app.jsx]: I'd prefer if we keep filenames to lowercase, e.g. `use-model` to avoid case-sensitive file system bugs and more aligned with the node eco system

File: example/app.jsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    (txt, utf8, embedded labels, delimiter is ',');`;
+    await app.setScript(script);
+    await app.doReload();
+    return app;
+  }, [global]);
+  return sessionApp;
+}
+
+export default function App() {
+  const element = useRef(null);
```
Comment (peol) [inline example/app.jsx]: Perhaps a comment on each of these (like `// we need to keep track of an element reference for the picasso chart`, or `// we use the useMemo hook to avoid creating new enigma.js sessions each time`)

File: example/app.jsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  const session = useMemo(() => enigma.create({ schema, url: 'ws://localhost:9076/app' }), [false]);
+  const [global] = useGlobal(session);
+  const app = useSessionApp(global);
+  const [model, modelError] = useModel(app, props);
+  const [layout, layoutError] = useLayout(model);
+  usePicasso(element, settings, layout);
```
Comment (peol) [inline example/app.jsx]: Can we do something with the picasso chart instance, `const pic = usePicasso(...)` so it's clear that they can interact with it from outside the hook?

File: example/data/tarantino.csv [hunk 9]
```diff
@@ -0,0 +1,1895 @@
+movie,type,word,minutes_in
```
Comment (peol) [inline example/data/tarantino.csv]: Where did we get this data from? We need to reference a license. Also, lots of swearwords etc. here... Perhaps we should avoid those datasets in examples? :)

File: package.json [hunk 10]
```diff
@@ -0,0 +1,60 @@
+{
+  "name": "hamus.js",
+  "version": "0.0.1",
+  "description": "Collections of React Hooks for use with QLIK APIs",
```
Comment (peol) [inline package.json]: Qlik

File: src/hooks/useLayout.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+export default function useLayout(model) {
+  const [changed, setChanged] = useState(null);
+  let canceled = false;
+  let layout = null;
+  let error = null;
+
+  [layout, error] = usePromise(() => {
+    if (model) {
+      const newLayout = model.getAppLayout ? model.getAppLayout() : model.getLayout();
+      if (!canceled) {
```
Comment (peol) [inline src/hooks/useLayout.js]: Maybe move this to the top, so `if (!model || canceled) return null`, then you can remove the last `return null` too

File: src/hooks/usePicasso.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import { useState, useEffect } from 'react';
+import picassoQ from 'picasso-plugin-q';
+import picasso from 'picasso.js';
+
+picasso.use(picassoQ);
```
Comment (peol) [inline src/hooks/usePicasso.js]: If we do this, we *might* affect the global instance of picasso (e.g. affect the users own configuration of picasso). Perhaps @miralemd has insights how we should do this.

File: LICENSE (hunk 1 above)
Comment (hrigner) [reply LICENSE]: 🙈

File: docs/useModel.md (hunk 3 above)
Comment (hrigner) [reply docs/useModel.md]: I agree! Pretty small package as well :)

File: src/hooks/usePicasso.js (hunk 12 above)
Comment (miralemd) [reply src/hooks/usePicasso.js]: it's possible to create an isolated instance of picasso:

```js
import picassoQ from 'picasso-plugin-q';
import picassojs from 'picasso.js';

const picasso = picassojs(); // create a new instance of picassojs to avoid affecting the global
picasso.use(picassoQ); // picassoQ will now only be available on the custom instance and not on the global

picasso.chart(/* ... */);
```

File: example/data/tarantino.csv (hunk 9 above)
Comment (hrigner) [reply example/data/tarantino.csv]: It's from fivethirtyeight, https://data.fivethirtyeight.com/ .

I was using it for another project and kept it for simplicity/laziness.

File: src/hooks/useLayout.js (hunk 11 above)
Comment (hrigner) [reply src/hooks/useLayout.js]: Then we miss the narrow case that canceled turns true during the call to getLayout.. and then we return the layout instead of null.

But maybe that's ok?

File: example/app.jsx (hunk 8 above)
Comment (hrigner) [reply example/app.jsx]: yepp, we can do some brushing. I did think of that, but out of laziness I accidentally forgot it ;)

File: src/hooks/usePicasso.js (hunk 12 above)
Comment (hrigner) [reply src/hooks/usePicasso.js]: Ah, sweet.

File: src/hooks/useLayout.js (hunk 11 above)
Comment (peol) [reply src/hooks/useLayout.js]: We're not waiting for the layout (no await) so it's synchronous :)

File: src/hooks/usePicasso.js (hunk 12 above)
Comment (peol) [reply src/hooks/usePicasso.js]: The tricky thing with that is that we then are using "our own" instance... Maybe that's OK, otherwise I think we should let the user pass in the picasso instancce

File: .babelrc [hunk 13]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (peol) [inline .babelrc]: Why is this needed, for the tests? Or the example?

File: .eslintrc.json [hunk 14]
```diff
@@ -0,0 +1,16 @@
+{
+  "env": {
+    "browser": true,
+    "jest": true
```
Comment (peol) [inline .eslintrc.json]: jest should only be enabled in `test/` or you might end up with test things in `src/`. You can inherit this config in `test/` like this: https://github.com/qlik-oss/enigma.js/blob/master/test/.eslintrc

File: docs/useLayout.md [hunk 15]
```diff
@@ -0,0 +1,40 @@
+# useLayout
+
+Fetches the layout from a model, and updates the layout on model changes.
+Calls either the QIX API method [getAppLayout](https://core.qlik.com/services/qix-engine/apis/qix/doc/#getapplayout) or [GetLayout](https://core.qlik.com/services/qix-engine/apis/qix/genericobject/#getlayout).
+
+## Usage
+
+```jsx
+import useLayout from 'hamus.js';
```
Comment (peol) [inline docs/useLayout.md]: Is this still true, or is it `import useLayout from 'hamus.js/use-layout';` now? I think it was better with the index file that was exporting them if so. Sorry for the confusion :)

File: package.json [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    "parcel-bundler": "1.12.0",
+    "picasso.js": "0.20.0",
+    "picasso-plugin-q": "0.20.0",
+    "react": "16.8.4",
+    "react-dom": "16.8.4"
+  },
+  "jest": {
+    "collectCoverage": true,
+    "collectCoverageOnlyFrom": {
+      "src/hooks/useLayout.js": true,
```
Comment (peol) [inline package.json]: These paths needs to be updated

File: .eslintrc.json (hunk 14 above)
Comment (miralemd) [reply .eslintrc.json]: or you could [override](https://eslint.org/docs/user-guide/configuring#disabling-rules-only-for-a-group-of-files) config for certain files to avoid having multiple eslint files in the project as we do in [picasso](https://github.com/qlik-oss/picasso.js/blob/master/.eslintrc.json#L33).

File: docs/useLayout.md (hunk 15 above)
Comment (hrigner) [reply docs/useLayout.md]: I'll go back to using the index file.

File: package.json (hunk 16 above)
Comment (hrigner) [reply package.json]: oops

File: .babelrc (hunk 13 above)
Comment (hrigner) [reply .babelrc]: Both of them..

File: example/app.jsx [hunk 17]
```diff
@@ -0,0 +1,143 @@
+import React, { useMemo, useRef } from 'react';
+import enigma from 'enigma.js';
+import schema from 'enigma.js/schemas/12.170.2.json';
+import picasso from 'picasso.js';
+import picassoQ from 'picasso-plugin-q';
+import usePromise from 'react-use-promise';
+import useModel from '../src/use-model';
```
Comment (peol) [inline example/app.jsx]: `import { useModel, useLayout, usePicasso } from '../src/index`;`

Comment (hrigner) [reply example/app.jsx]: yepp!
